File: component_demo1/src/main/java/com/example/component_demo1/http/HttpCallback.java
package com.example.component_demo1.http;

import android.content.Context;
import com.example.base_fun.http.LibHttpCallback;
import org.json.JSONException;
import retrofit2.Call;
import retrofit2.Response;

public abstract class HttpCallback<T> extends LibHttpCallback {
    private Context context;

    public HttpCallback() {
    }

    public HttpCallback(String serviceName) {
        super(serviceName);
    }

    public HttpCallback(Context context, boolean showProgress) {
        this.context = context;
    }

    public HttpCallback(Context context, boolean showProgress, String serviceName) {
        super(serviceName);
        this.context = context;
    }

    @Override
    public void onResponse(Call call, Response response) {
        if (call.isCanceled()) {
            onCancel();
            return;
        }

        if (response == null || response.body() == null || (!response.isSuccessful())) {
            onFaile(new Exception("BNET404"));//网络请求异常
            return;
        }

        T t = null;
        try {
            t = getBean(response);
        } catch (Exception e) {
            t = null;
        }


        if (t == null) {
            onFaile(new Exception("net_inner_method")); //内部方法出错
            return;
        }
//            t.setServiceName(getServiceName());
//            if (t.isSuccess()) {
//                onSucess(t);
//                return;
//            }

        onFaile(new Exception("")); //服务器返回的错误
    }

    @Override
    public void onFailure(Call call, Throwable t) {
        if (call.isCanceled()) {
            onCancel();
            return;
        }

        onFaile(new Exception(t));
    }

    public void onSucess(T t) {
        dismiss();
    }

    public void onFaile(Exception ex) {
        dismiss();
    }

    public void onCancel() {
        dismiss();
    }

    protected T getBean(Response response) throws JSONException {
//        if (getType() == Companion.getTypeGson()) {
//            return ((Resp<T>) response.body()).getResponse();
//        } else if (getType() == Companion.getTypeXml()) {
//            return (T) response.body();
//        } else if (getType() == Companion.getTypeString()) {//需要重写方法
//        }

        return null;
    }

    public HttpCallback<T> setServiceName(String serviceName) {
        super.setServiceNameString(serviceName);
        return this;
    }

    public HttpCallback<T> asyn(boolean asyn) {
        super.setAsyn(asyn);
        return this;
    }

    public Context getContext() {
        return context;
    }

    public void dismiss() {
    }
}

File: component_demo2/src/main/java/com/example/component_demo2/http/HttpCallback.kt
package com.example.component_demo2.http

import android.content.Context
import com.example.base_fun.http.LibHttpCallback
import org.json.JSONException
import retrofit2.Call
import retrofit2.Response

abstract class HttpCallback<T> : LibHttpCallback {
    var context: Context? = null
        private set

    constructor()

    constructor(serviceName: String) : super(serviceName)

    constructor(context: Context, showProgress: Boolean) {
        this.context = context
    }

    constructor(context: Context, showProgress: Boolean, serviceName: String) : super(serviceName) {
        this.context = context
    }

    final override fun onResponse(call: Call<Any>, response: Response<Any>) {
        if (call.isCanceled) {
            onCancel()
            return
        }

        if (response?.body() == null || !response.isSuccessful) {
            onFaile(Exception("BNET404"))//网络请求异常
            return
        }

        var t: T? = null
        try {
            t = getBean(response)
        } catch (e: Exception) {
            t = null
        }


        if (t == null) {
            onFaile(Exception("net_inner_method")) //内部方法出错
            return
        }
        //            t.setServiceName(getServiceName());
        //            if (t.isSuccess()) {
        //                onSucess(t);
        //                return;
        //            }

        onFaile(Exception("")) //服务器返回的错误
    }

    final override fun onFailure(call: Call<Any>, t: Throwable) {
        if (call.isCanceled) {
            onCancel()
            return
        }

        onFaile(Exception(t))
    }

    open fun onSucess(t: T) {
        dismiss()
    }

    open fun onFaile(ex: Exception) {
        dismiss()
    }

    open fun onCancel() {
        dismiss()
    }

    @Throws(JSONException::class)
    protected open fun getBean(response: Response<*>): T? {
        //        if (getType() == Companion.getTypeGson()) {
        //            return ((Resp<T>) response.body()).getResponse();
        //        } else if (getType() == Companion.getTypeXml()) {
        //            return (T) response.body();
        //        } else if (getType() == Companion.getTypeString()) {//需要重写方法
        //        }

        return null
    }

    fun setServiceName(serviceName: String): HttpCallback<T> {
        super.setServiceNameString(serviceName)
        return this
    }

    fun asyn(asyn: Boolean): HttpCallback<T> {
        super.isAsyn = asyn
        return this
    }

    fun dismiss() {}
}